cli/client.go

```diff
@@ -0,0 +1,305 @@
+	Uptime        time.Time         `json:"uptime"`
+	ClusterStatus *ClusterStatus    `json:"clusterStatus"`
+}
+
+// PeerStatus represents the status of a peer in the cluster.
+type PeerStatus struct {
+	Name    string `json:"name"`
+	Address string `json:"address"`
+}
+
+// ClusterStatus represents the status of the cluster.
+type ClusterStatus struct {
+	Name   string       `json:"name"`
+	Status string       `json:"status"`
+	Peers  []PeerStatus `json:"peers"`
+}
+
+// apiClient wraps a regular client and processes successful API responses.
+// Successful also includes responses that errored at the API level.
+type apiClient struct {
+	api.Client
+}
+
+type apiResponse struct {
+	Status    string          `json:"status"`
+	Data      json.RawMessage `json:"data,omitempty"`
+	ErrorType string          `json:"errorType,omitempty"`
+	Error     string          `json:"error,omitempty"`
+}
+
+type clientError struct {
+	code int
+	msg  string
+}
+
+func (e *clientError) Error() string {
+	return fmt.Sprintf("%s (code: %d)", e.msg, e.code)
+}
+
+func (c apiClient) Do(ctx context.Context, req *http.Request) (*http.Response, []byte, error) {
+	resp, body, err := c.Client.Do(ctx, req)
+	if err != nil {
+		return resp, body, err
+	}
+
+	code := resp.StatusCode
+
+	var result apiResponse
+	if err = json.Unmarshal(body, &result); err != nil {
+		// Pass the returned body rather than the JSON error because some API
+		// endpoints return plain text instead of JSON payload.
+		return resp, body, &clientError{
+			code: code,
+			msg:  string(body),
+		}
+	}
+
+	if (code/100 == 2) && (result.Status != statusSuccess) {
+		return resp, body, &clientError{
+			code: code,
+			msg:  "inconsistent body for response code",
+		}
+	}
+
+	if result.Status == statusError {
+		err = &clientError{
+			code: code,
+			msg:  result.Error,
+		}
+	}
+
+	return resp, []byte(result.Data), err
+}
+
+// StatusAPI provides bindings for the Alertmanager's status API.
+type StatusAPI interface {
+	// Get returns the server's configuration, version, uptime and cluster information.
+	Get(ctx context.Context) (*ServerStatus, error)
+}
+
+// NewStatusAPI returns a status API client.
+func NewStatusAPI(c api.Client) StatusAPI {
+	return &httpStatusAPI{client: apiClient{c}}
+}
+
+type httpStatusAPI struct {
+	client api.Client
+}
+
+func (h *httpStatusAPI) Get(ctx context.Context) (*ServerStatus, error) {
+	u := h.client.URL(epStatus, nil)
+
+	req, _ := http.NewRequest(http.MethodGet, u.String(), nil)
+
+	_, body, err := h.client.Do(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+
+	var ss *ServerStatus
+	err = json.Unmarshal(body, &ss)
+
+	return ss, err
+}
+
+// AlertAPI provides bindings for the Alertmanager's alert API.
+type AlertAPI interface {
+	// List returns all the active alerts.
+	List(ctx context.Context) ([]*model.Alert, error)
+	// Push sends a list of alerts to the Alertmanager.
+	Push(ctx context.Context, alerts ...model.Alert) error
```

Comment:
I'm having import issues when trying to use this library, since there's an import mismatch:

```
./main.go:50:15: cannot use alert (type "github.com/prometheus/common/model".Alert) as type "github.com/prometheus/alertmanager/vendor/github.com/prometheus/common/model".Alert in argument to alertAPI.Push
```

We're running into issue because alertmanager is being used as a library, which shouldn't vendor this, when it was originally an application, which should vendor. We might have to introduce an intermediate type for the API.

Reply:
Still having issues with this. The easiest thing to do is to introduce the intermediate type, and then in a separate PR we can decide if we want to replace references to `model.Alert` throughout the code base.

@brian-brazil what's the current state on using `common/model`? Are we deprecating it?

Reply:
I had success when copying over the types from `common/model`:
```diff
diff --git a/cli/client.go b/cli/client.go
index c623b25..9d7e0d4 100644
--- a/cli/client.go
+++ b/cli/client.go
@@ -25,7 +25,6 @@ import (
 
 	"github.com/prometheus/alertmanager/config"
 	"github.com/prometheus/alertmanager/types"
-	"github.com/prometheus/common/model"
 )
 
 const (
@@ -154,9 +153,9 @@ func (h *httpStatusAPI) Get(ctx context.Context) (*ServerStatus, error) {
 // AlertAPI provides bindings for the Alertmanager's alert API.
 type AlertAPI interface {
 	// List returns all the active alerts.
-	List(ctx context.Context) ([]*model.Alert, error)
+	List(ctx context.Context) ([]*Alert, error)
 	// Push sends a list of alerts to the Alertmanager.
-	Push(ctx context.Context, alerts ...model.Alert) error
+	Push(ctx context.Context, alerts ...Alert) error
 }
 
 // NewAlertAPI returns a new AlertAPI for the client.
@@ -168,7 +167,20 @@ type httpAlertAPI struct {
 	client api.Client
 }
 
-func (h *httpAlertAPI) List(ctx context.Context) ([]*model.Alert, error) {
+type Alert struct {
+	Labels       LabelSet  `json:"labels"`
+	Annotations  LabelSet  `json:"annotations"`
+	StartsAt     time.Time `json:"startsAt,omitempty"`
+	EndsAt       time.Time `json:"endsAt,omitempty"`
+	GeneratorURL string    `json:"generatorURL"`
+}
+
+type LabelSet map[LabelName]LabelValue
+type LabelName string
+type LabelValue string
+
+// List implements the AlertAPI interface.
+func (h *httpAlertAPI) List(ctx context.Context) ([]*Alert, error) {
 	u := h.client.URL(epAlerts, nil)
 
 	req, _ := http.NewRequest(http.MethodGet, u.String(), nil)
@@ -178,13 +190,14 @@ func (h *httpAlertAPI) List(ctx context.Context) ([]*model.Alert, error) {
 		return nil, err
 	}
 
-	var alts []*model.Alert
-	err = json.Unmarshal(body, &alts)
+	var alerts []*Alert
+	err = json.Unmarshal(body, &alerts)
 
-	return alts, err
+	return alerts, err
 }
 
-func (h *httpAlertAPI) Push(ctx context.Context, alerts ...model.Alert) error {
+// Push implements the AlertAPI interface.
+func (h *httpAlertAPI) Push(ctx context.Context, alerts ...Alert) error {
 	u := h.client.URL(epAlerts, nil)
 
 	var buf bytes.Buffer
```

Reply:
> @brian-brazil what's the current state on using common/model? Are we deprecating it?

The general direction is to replace it, but there's still going to be a common library one way or another.